File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n = matrix.size();
        vector<vector<int>> dp(n, vector<int>(n, 0));
        dp[0] = matrix[0];
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < n; j++) {
                int temp = dp[i - 1][j];
                if (j > 0) {
                    temp = min(temp, dp[i - 1][j - 1]);
                }
                if (j < n - 1) {
                    temp = min(temp, dp[i - 1][j + 1]);
                }
                dp[i][j] = matrix[i][j] + temp;
            }
        }
        return *min_element(dp.back().begin(), dp.back().end());
    }
};

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int m = matrix.length, n = matrix[0].length;
        int result = Integer.MAX_VALUE;
        for (int i = 1; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int topLeft = matrix[i - 1][Math.max(j - 1, 0)];
                int topRight = matrix[i - 1][Math.min(j + 1, n - 1)];
                int top = matrix[i - 1][j];
                matrix[i][j] += Math.min(top, Math.min(topLeft, topRight));
            }
        }
        for (int j = 0; j < n; j++) {
            result = Math.min(result, matrix[m - 1][j]);
        }
        return result;
    }
}